node/network/collator-protocol/src/validator_side.rs

```diff
@@ -0,0 +1,1360 @@
+use polkadot_node_network_protocol::{
+	v1 as protocol_v1, View, PeerId, ReputationChange as Rep, RequestId,
+	NetworkBridgeEvent,
+};
+
+use super::{modify_reputation, TARGET, Result};
+
+const COST_UNEXPECTED_MESSAGE: Rep = Rep::new(-10, "An unexpected message");
+const COST_REQUEST_TIMED_OUT: Rep = Rep::new(-20, "A collation request has timed out");
+const COST_REPORT_BAD: Rep = Rep::new(-50, "A collator was reported by another subsystem");
+const BENEFIT_NOTIFY_GOOD: Rep = Rep::new(50, "A collator was noted good by another subsystem");
+
+#[derive(Debug)]
+enum CollationRequestResult {
+	Received(RequestId),
+	Timeout(RequestId),
+}
+
+/// A Future representing an ongoing collation request.
+/// It may timeout or end in a graceful fashion if a requested
+/// collation has been received sucessfully or chain has moved on.
+struct CollationRequest {
+	// The response for this request has been received successfully or 
+	// chain has moved forward and this request is no longer relevant.
+	received: oneshot::Receiver<()>,
+
+	// The timeout of this request.
+	timeout: Delay,
+
+	// The id of this request.
+	request_id: RequestId,
+}
+
+impl Future for CollationRequest {
+	type Output = CollationRequestResult;
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll::<Self::Output> {
+		match Pin::new(&mut self.timeout).poll(cx) {
+			Poll::Pending => (),
+			Poll::Ready(_) => {
+				return Poll::Ready(CollationRequestResult::Timeout(self.request_id));
+			}
+		};
+
+		match Pin::new(&mut self.received).poll(cx) {
+			Poll::Pending => (),
+			Poll::Ready(_) => {
+				return Poll::Ready(CollationRequestResult::Received(self.request_id));
+			}
+		};
+
+		Poll::Pending
+	}
+}
+
+struct PerRequest {
+	// The sender side to signal the `CollationRequest` to resolve successfully.
+	received: oneshot::Sender<()>,
+
+	// Send result here.
+	// `Some` in case the collation was requested by some other subsystem.
+	// `None` in case the collation was requested by this subsystem.
+	result: Option<mpsc::Sender<(CollatorId, CandidateReceipt, PoV)>>,
+}
+
+/// All state relevant for the validator side of the protocol lives here.
+#[derive(Default)]
+struct State {
+	/// Our own view.
+	view: View,
+
+	/// Track all active collators and their views.
+	peer_views: HashMap<PeerId, View>,
+
+	/// Peers that have declared themselves as collators.
+	known_collators: HashMap<PeerId, CollatorId>,
+
+	/// Advertisments received from collators. We accept one advertisment
+	/// per collator per source per relay-parent.
+	advertisments: HashMap<PeerId, HashSet<(ParaId, Hash)>>,
+
+	/// Derive RequestIds from this.
+	next_request_id: RequestId,
+
+	/// The collations we have requested by relay parent and para id.
+	///
+	/// For each relay parent and para id we may be connected to a number
+	/// of collators each of those may have advertised a different collation.
+	/// So we group such cases here.
+	requested_collations: HashMap<(Hash, ParaId, PeerId), RequestId>,
```

Comment:
Also worth noting that we don't need to fetch any of the collations by collators.